// @sanity/structure/lib/GenericList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericListBuilder = void 0;

var _camelCase2 = _interopRequireDefault(require("lodash/camelCase"));

var _Layout = require("./Layout");

var _MenuItem = require("./MenuItem");

var _MenuItemGroup = require("./MenuItemGroup");

var _SerializeError = require("./SerializeError");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function noChildResolver() {
  return undefined;
}

class GenericListBuilder {
  constructor() {
    _defineProperty(this, "spec", {});
  }

  id(id) {
    return this.clone({
      id
    });
  }

  getId() {
    return this.spec.id;
  }

  title(title) {
    return this.clone({
      title,
      id: this.spec.id || (0, _camelCase2.default)(title)
    });
  }

  getTitle() {
    return this.spec.title;
  }

  defaultLayout(defaultLayout) {
    return this.clone({
      defaultLayout
    });
  }

  getDefaultLayout() {
    return this.spec.defaultLayout;
  }

  menuItems(menuItems) {
    return this.clone({
      menuItems
    });
  }

  getMenuItems() {
    return this.spec.menuItems;
  }

  menuItemGroups(menuItemGroups) {
    return this.clone({
      menuItemGroups
    });
  }

  getMenuItemGroups() {
    return this.spec.menuItemGroups;
  }

  child(child) {
    return this.clone({
      child
    });
  }

  getChild() {
    return this.spec.child;
  }

  canHandleIntent(canHandleIntent) {
    return this.clone({
      canHandleIntent
    });
  }

  getCanHandleIntent() {
    return this.spec.canHandleIntent;
  }

  showIcons(enabled) {
    return this.clone({
      displayOptions: _objectSpread({}, this.spec.displayOptions || {}, {
        showIcons: enabled
      })
    });
  }

  getShowIcons() {
    return this.spec.displayOptions ? this.spec.displayOptions.showIcons : undefined;
  }

  serialize() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    var id = this.spec.id || '';
    var path = options.path;
    var defaultLayout = this.spec.defaultLayout;

    if (defaultLayout && !_Layout.layoutOptions.includes(defaultLayout)) {
      throw new _SerializeError.SerializeError("`layout` must be one of ".concat(_Layout.layoutOptions.map(item => "\"".concat(item, "\"")).join(', ')), path, id || options.index, this.spec.title);
    }

    return {
      id,
      title: this.spec.title,
      type: 'genericList',
      defaultLayout,
      child: this.spec.child || noChildResolver,
      canHandleIntent: this.spec.canHandleIntent,
      displayOptions: this.spec.displayOptions,
      menuItems: (this.spec.menuItems || []).map((item, i) => (0, _MenuItem.maybeSerializeMenuItem)(item, i, path)),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) => (0, _MenuItemGroup.maybeSerializeMenuItemGroup)(item, i, path))
    };
  }

  clone(withSpec) {
    var builder = new this.constructor();
    return builder;
  }

}

exports.GenericListBuilder = GenericListBuilder;
